Database/Repositories/Implementations/TournamentsRepository.cs

```diff
@@ -90,6 +90,42 @@ public async Task<ICollection<Tournament>> GetAsync(int page, int pageSize, Tour
                 .ToListAsync();
     }
 
+    public async Task<ICollection<Beatmap>> AddPooledBeatmapsAsync(int id, ICollection<long> osuBeatmapIds)
+    {
+        Tournament? tournament = await _context.Tournaments
+            .Include(t => t.PooledBeatmaps)
+            .FirstOrDefaultAsync(t => t.Id == id);
+
+        if (tournament is null)
+        {
+            return new List<Beatmap>();
+        }
+
+        // Fetch the non-existent beatmap ids
+        var uniqueBeatmapIds = osuBeatmapIds.Except(await _context.Beatmaps
+            .Where(b => osuBeatmapIds.Contains(b.OsuId))
+            .Select(b => b.OsuId)
+            .ToListAsync()).ToList();
+
+        // Insert new beatmaps
+        IEnumerable<Beatmap> newBeatmaps = uniqueBeatmapIds.Select(osuId => new Beatmap { OsuId = osuId });
+        await _context.Beatmaps.AddRangeAsync(newBeatmaps);
+        await _context.SaveChangesAsync();
+
+        // Gather all beatmaps which match this beatmap id collection and add to pooled beatmaps
+        List<Beatmap> beatmaps = await _context.Beatmaps.Where(b => osuBeatmapIds.Contains(b.OsuId)).ToListAsync();
+        beatmaps.ForEach(b =>
+        {
+            if (!tournament.PooledBeatmaps.Contains(b))
+            {
+                tournament.PooledBeatmaps.Add(b);
+            }
+        });
+
+        await UpdateAsync(tournament);
+        return tournament.PooledBeatmaps;
+    }
```

Comment:
I count 4 db calls in this function alone. A lot of your logic with finding and creating unique beatmaps is really unnecessary [1]. In fact, I think this function should probably completely move to the service instead. I'd recommend looking at the way I tweaked your implementation of this on `TournamentsService.CreateAsync(TournamentSubmissionDTO submission, int submitterUserId, bool preApprove)`. You could probably abstract that out into a private function that can be used for both the existing function and this new one.

[1] This comes off a bit abrasive so I'll explain:
The way you find non-existing beatmap ids is fine, what I have an issue mostly with is the way you create the new beatmaps. There is no reason to create, add, and then save these new beatmaps on their own when you can create them by proxy. If you add the newly created beatmap objects to the `Tournament.PooledBeatmaps` collection, they will be saved to the database on the next save by proxy via EF :D This is exactly the same way we do it with matches and beatmaps in the service method

Reply:
I don't see a reason for this being moved to the service. This is change is suited more for the repository class in my eyes. I do agree that it's inefficient so I'll fix that.

Reply:
I see where you're coming from but I'm just thinking in terms of consistency. The `TournamentsService.CreateAsync()` method does the exact same thing for creating beatmaps but on the service level instead. Do you think that method should be moved down to the repository as well?

Reply:
Opting to add this functionality to `TournamentsRepository` instead as it feels very odd to have to write a dedicated `GetAsync` function for including a specific navigation (`PooledBeatmaps` in this case).